lib/benchmarks/src/Cardano/Wallet/Benchmarks/History.hs

```diff
@@ -0,0 +1,187 @@
+import qualified Data.Text as T
+import qualified Data.Vector as V
+
+data IndexedSemantic = IndexedSemantic
+    { semantic :: Semantic
+    , index :: Int
+    }
+    deriving stock (Eq, Ord, Show)
+
+type History = MonoidalMap IndexedSemantic (MonoidalMap Day (First Result))
+
+data HarmonizedRow
+    = Harmonized (Map Day Double) Units
+    | FailedToHarmonize [(Day, Result)]
+    deriving stock (Show)
+
+harmonizeUnits :: [(Day, Result)] -> HarmonizedRow
+harmonizeUnits rs =
+    let
+        u = biggestUnit $ fmap (resultUnits . snd) rs
+    in
+        case u of
+            Nothing -> FailedToHarmonize rs
+            Just u' ->
+                Harmonized
+                    (Map.fromList $ second (harmonizeResult u') <$> rs)
+                    u'
+
+-- dropping the iteration count ... for now
+harmonizeResult :: Units -> Result -> Double
+harmonizeResult u (Result v u' _i) = convertUnits u' u v
+
+biggestUnit :: [Units] -> Maybe Units
+biggestUnit = foldr go Nothing
+  where
+    go :: Units -> Maybe Units -> Maybe Units
+    go u Nothing = Just u
+    go u (Just u') = Just $ compareUnits u u'
+
+compareUnits :: Units -> Units -> Units
+compareUnits u u' = case (u, u') of
+    (Seconds, Seconds) -> Seconds
+    (Seconds, Milliseconds) -> Seconds
+    (Seconds, Microseconds) -> Seconds
+    (Seconds, Nanoseconds) -> Seconds
+    (Milliseconds, Seconds) -> Seconds
+    (Milliseconds, Milliseconds) -> Milliseconds
+    (Milliseconds, Microseconds) -> Milliseconds
+    (Milliseconds, Nanoseconds) -> Milliseconds
+    (Microseconds, Seconds) -> Seconds
+    (Microseconds, Milliseconds) -> Milliseconds
+    (Microseconds, Microseconds) -> Microseconds
+    (Microseconds, Nanoseconds) -> Microseconds
+    (Nanoseconds, Seconds) -> Seconds
+    (Nanoseconds, Milliseconds) -> Milliseconds
+    (Nanoseconds, Microseconds) -> Microseconds
+    (Nanoseconds, Nanoseconds) -> Nanoseconds
+    (Bytes, Bytes) -> Bytes
+    (Bytes, KiloBytes) -> KiloBytes
+    (Bytes, MegaBytes) -> MegaBytes
+    (Bytes, GigaBytes) -> GigaBytes
+    (KiloBytes, Bytes) -> KiloBytes
+    (KiloBytes, KiloBytes) -> KiloBytes
+    (KiloBytes, MegaBytes) -> MegaBytes
+    (KiloBytes, GigaBytes) -> GigaBytes
+    (MegaBytes, Bytes) -> MegaBytes
+    (MegaBytes, KiloBytes) -> MegaBytes
+    (MegaBytes, MegaBytes) -> MegaBytes
+    (MegaBytes, GigaBytes) -> GigaBytes
+    (GigaBytes, Bytes) -> GigaBytes
+    (GigaBytes, KiloBytes) -> GigaBytes
+    (GigaBytes, MegaBytes) -> GigaBytes
+    (GigaBytes, GigaBytes) -> GigaBytes
+    (Count, Count) -> Count
+    _ -> error "Invalid units"
+
+data HarmonizedHistory = HarmonizedHistory
+    { harmonizedHistory :: MonoidalMap IndexedSemantic HarmonizedRow
+    , days :: Set Day
+    }
+    deriving stock (Show)
+
+harmonizeDay :: MonoidalMap Day (First Result) -> HarmonizedRow
+harmonizeDay = harmonizeUnits . fmap (second getFirst) . assocs
+
+harmonizeHistory :: History -> HarmonizedHistory
```

Comment:
```suggestion
harmonizeHistory :: History -> Either ErrHarmonize HarmonizedHistory
```

I have a feeling that moving the error case to the top of the data structure makes the definition of `renderHarmonizedHistoryCsv ` simpler — at the moment, that definition fails if a single row fails to harmonize anyway.